recorder/src/main/cpp/site_resolver.cc

```diff
@@ -0,0 +1,107 @@
+#include "site_resolver.hh"
+
+bool SiteResolver::method_info(const jmethodID method_id, jvmtiEnv* jvmti, MethodListener& listener) {
+    jint error;
+    JvmtiScopedPtr<char> methodName(jvmti);
+    JvmtiScopedPtr<char> methodSig(jvmti);
+
+    error = jvmti->GetMethodName(method_id, methodName.GetRef(), methodSig.GetRef(), NULL);
+    if (error != JVMTI_ERROR_NONE) {
+        methodName.AbandonBecauseOfError();
+        methodSig.AbandonBecauseOfError();
+        if (error == JVMTI_ERROR_INVALID_METHODID) {
+            static int once = 0;
+            if (!once) {
+                once = 1;
+                logError("One of your monitoring interfaces "
+                         "is having trouble resolving its stack traces.  "
+                         "GetMethodName on a jmethodID involved in a stacktrace "
+                         "resulted in an INVALID_METHODID error which usually "
+                         "indicates its declaring class has been unloaded.\n");
+                logError("Unexpected JVMTI error %d in GetMethodName\n", error);
+            }
+        }
+        return false;
+    }
+
+    // Get class name, put it in signature_ptr
+    jclass declaring_class;
+    JVMTI_ERROR_RET(
+        jvmti->GetMethodDeclaringClass(method_id, &declaring_class), false);
+
+    JvmtiScopedPtr<char> signature_ptr2(jvmti);
+    JVMTI_ERROR_CLEANUP_RET(
+        jvmti->GetClassSignature(declaring_class, signature_ptr2.GetRef(), NULL),
+        false, signature_ptr2.AbandonBecauseOfError());
+
+    // Get source file, put it in source_name_ptr
+    char *fileName;
+    JvmtiScopedPtr<char> source_name_ptr(jvmti);
+    static char file_unknown[] = "UnknownFile";
+    if (JVMTI_ERROR_NONE !=
+        jvmti->GetSourceFileName(declaring_class, source_name_ptr.GetRef())) {
+        source_name_ptr.AbandonBecauseOfError();
+        fileName = file_unknown;
+    } else {
+        fileName = source_name_ptr.Get();
+    }
+
+    listener.recordNewMethod(method_id, fileName, signature_ptr2.Get(), methodName.Get(), methodSig.Get());
+
+    return true;
+}
+
+static jint bci2line(jint bci, jvmtiLineNumberEntry *table, jint entry_count) {
+	jint line_number = -101;
+	if ( entry_count == 0 ) {
+		return line_number;
+	}
+	line_number = -102;
```

Comment:
This is old honest-profiler code just moved around and re-organized. I haven't really tried to re-do it yet.